File: backend/src/main/java/com/main/Difficult/Domain/producto/Producto.java
package com.main.Difficult.Domain.producto;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonView;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.HashSet;
import java.util.Set;

@JsonView(View.ProductoLista.class)
@Document("Producto")
@Getter @Setter
public abstract class Producto {
    @Id
    @JsonView(View.ProductoRecomendado.class) private String id;
    @JsonView(View.ProductoRecomendado.class) private String imagen;
    @JsonView(View.ProductoRecomendado.class) private String nombre;
    @JsonView(View.ProductoRecomendado.class) private String descripcion;
    private Integer puntaje = 0;
    private String paisDeOrigen = "";
    @JsonIgnore  private Double precioBase = 0.0;
    @JsonProperty @JsonView(View.ProductoConLote.class)
    private Set<Lote> lotes = new HashSet<>() ;

    public Producto(String imagen, String nombre, String descripcion, Integer puntaje, String paisDeOrigen, Double precioBase){
        this.setImagen(imagen);
        this.setNombre(nombre);
        this.setDescripcion(descripcion);
        this.setPuntaje(puntaje);
        this.setPaisDeOrigen(paisDeOrigen);
        this.setPrecioBase(precioBase);
    }

    public Lote lotePorId(Long loteId){
        return lotes.stream()
                .filter(lote->lote.getId().equals(loteId))
                .findAny()
                .orElse(null);
    }
    public void agregarLote(Lote lote){
        lotes.add(lote);
    }
    @JsonProperty
    @JsonView(View.ProductoRecomendado.class)
    public Double precio(){
        return precioBase * this.descuento();
    }
    public Double descuento(){
        return this.hayProductoConMasDe4Meses() ?  0.90 : 1;
    }
    public Boolean hayProductoConMasDe4Meses(){
        return lotes.stream().anyMatch(lote-> lote.tieneMasDe4Meses());
    }
    @JsonProperty public String tipo(){
        return this.getClass().getSimpleName();
    }

}

File: backend/src/main/kotlin/aplicacion/dominio/producto/Producto.kt
package aplicacion.dominio.producto

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonView
import javax.persistence.*

@JsonView(View.ProductoLista::class)
@Entity @Inheritance
abstract class Producto {
    @Id @GeneratedValue(strategy=GenerationType.IDENTITY)
    open var id: Long? = null
    open var imagen: String = ""

    open var nombre: String = ""
    open var descripcion: String = ""
    open var puntaje: Int = 0
    open var paisDeOrigen: String = ""

    @JsonIgnore
    open var precioBase: Double = 0.0


    @JsonView(View.ProductoIndividual::class)
    @OneToMany(cascade = [CascadeType.MERGE,CascadeType.PERSIST]) @JoinColumn(name="producto_id")
    open var lotes: MutableSet<Lote> = mutableSetOf()

    fun agregarLote(lote: Lote) {
        lotes.add(lote)
    }

    fun quitarLote(lote: Lote) {
        lotes.remove(lote)
    }

    fun lotePorId(liteId: Long): Lote {
        return lotes.first { it.id == liteId }
    }

    @JsonProperty
    open fun precio(): Double {
        return precioBase * descuento()
    }

    open fun descuento(): Double {
        return if (hayProductoConMasDe4Meses()) 0.90 else 1.00
    }


    open fun hayProductoConMasDe4Meses(): Boolean {
        return lotes.any { it.tieneMasDe4Meses() }
    }

    @JsonProperty
    fun tipo(): String? {
        return this::class.simpleName
    }
}